### DISCUSSION TITLE
One table schema in multiple databases?

### DISCUSSION THREAD

**Discussion by sqcsabbey** [simolus3/drift#1981] [Q&A] [answered]
When I try to use the same table definition in multiple databases, the DataClass for it is getting generated multiple times, which doesn't work...

Why would I want to do this? I want to use separate databases for easy export/import of particular data, and having the same table definition makes it easy to insert/update in multiple places. 

Is there a way to make this work without a lot of additional code?

**Comment by simolus3** [comment, marked as answer]
By far, the easiest and cleanest way to do this is to just let drift generate one database class.

You could do something like this for instance:

```dart
// This is the one that drift generates
@DriftDatabase(...)
abstract class BaseDatabase extends _$BaseDatabase {
  BaseDatabase(super.executor);
}

// You could then write your own independent database classes like this:
class MyDatabase1 extends BaseDatabase {
  MyDatabase1(super.executor);

  // override schemaVersion and so on...
}
class MyDatabase2 extends BaseDatabase {
  MyDatabase2(super.executor);

  // override schemaVersion and so on...
}
```

Would that work for you?

If the databases are completely identical and you only need more than one because they're storing data in different files, I would just use one database class and open that multiple times. You can safely set `driftRuntimeOptions.dontWarnAboutMultipleDatabases = true`in that case.

**Comment by sqcsabbey** [comment]
Thank you! That could work, but I'd have a lot of empty tables. 

My plan was to use one main database with a projects table, and then multiple project databases also with a projects table, but only with a single row populated, for itself. This would let me share just a project database file, and then the project name and other details could be imported into the main database. ... this was going to be the only table in common between the 2 database schemas. 

Any chance this could be supported? From my naive vantage point, it looks like the DataClass just shouldn't be generated if it already exists. 

And thanks for the `dontWarnAboutMultipleDatabases` option; that is really useful.

**Reply by simolus3** [reply]
If this is just about imports and exports, I'd probably create a new database file during each export and extract the relevant tables / rows during each export step.

> Any chance this could be supported? From my naive vantage point, it looks like the DataClass just shouldn't be generated if it already exists.

It's not that easy since all drift builders run concurrently, so there's no well-defined way of defining where the data class should go to. In a future drift version, there will hopefully be a better solution to this (#490). One thing that comes to mind now would be:

1. To split your build into two [targets](https://pub.dev/packages/build_config#dividing-a-package-into-build-targets). One target that only includes the main database and a second target for all other databases. The second target should have a dependency on the first one.
2. You can then write a second table class that extends the first one, but uses the generated data class from the first target as an [existing row class](https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/) with an annotation. This is the project table you'd use on the secondary databases.
